src/tools/pkcs11-tool.c

```diff
@@ -3913,6 +3913,55 @@ parse_ec_pkey(EVP_PKEY *pkey, int private, struct gostkey_info *gost)
 
 	return 0;
 }
+static int
+parse_ed_pkey(EVP_PKEY *pkey, int pk_type, int private, struct gostkey_info *gost)
+{
+	static unsigned char ec_params_ed25519[] = { 0x06, 0x03, 0x2b, 0x65, 0x70 };
+	static unsigned char ec_params_ed448[] = { 0x06, 0x03, 0x2b, 0x65, 0x71 };
+	unsigned char *ec_params = (pk_type == EVP_PKEY_ED25519) ? ec_params_ed25519 : ec_params_ed448;
+	size_t ec_params_size = (pk_type == EVP_PKEY_ED25519) ? sizeof(ec_params_ed25519) : sizeof(ec_params_ed448);
+	unsigned char *key;
+	size_t key_size;
+
+	/* set EC_PARAMS value */
+	gost->param_oid.value = malloc(ec_params_size);
+	if (gost->param_oid.value == NULL) {
+		return -1;
+	}
+	gost->param_oid.len = ec_params_size;
+	memcpy(gost->param_oid.value, ec_params, ec_params_size);
+
+	if (private) {
+		if (EVP_PKEY_get_raw_private_key(pkey, NULL, &key_size) != 1) {
+			return -1;
+		}
+	} else {
+		if (EVP_PKEY_get_raw_public_key(pkey, NULL, &key_size) != 1) {
+			return -1;
+		}
+	}
+
+	key = malloc(key_size);
+	if (key == NULL) {
+		return -1;
+	}
+
+	if (private) {
+		if (EVP_PKEY_get_raw_private_key(pkey, key, &key_size) != 1) {
+			return -1;
```

Comment:
free the `key` here otherwise you will leak the memory. The same in the following case

Reply:
I can do that. I'll be honest, I didn't quite know what the rules were given this is a command line tool which dies after it's done its work.